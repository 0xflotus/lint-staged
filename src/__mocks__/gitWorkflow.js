'use strict'

const hasPartiallyStagedFiles = jest.fn().mockImplementation(() => Promise.resolve(true))
const gitStashSave = jest.fn().mockImplementation(() => Promise.resolve(null))
const gitStashPop = jest.fn().mockImplementation(() => Promise.resolve(null))
const updateStash = jest.fn().mockImplementation(() => Promise.resolve(null))

module.exports = {
  gitStashSave,
  gitStashPop,
  updateStash,
  hasPartiallyStagedFiles
}
