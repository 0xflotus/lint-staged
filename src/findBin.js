'use strict'

const npmWhich = require('npm-which')(process.cwd())

module.exports = function findBin (binName, paths, config) {
    const bin = 'npm'
    const files = ['--color', '--'].concat(paths)
    const args = ['run', '-s', binName].concat(files)
    /*
    * If package.json has script with binName defined
    * we want it to be executed first
    */
    if (config.scripts[binName] !== undefined) {
        // Support for scripts from package.json
        return {
            bin,
            args
        }
    }

    /*
    *  If binName wasn't found in package.json scripts
    *  we'll try to locate the binary in node_modules/.bin
    *  This is useful for shorter configs like:
    *
    *  "lint-staged": {
    *    "*.js": "eslint"
    *  }
    *
    *  without adding
    *
    *  "scripts" {
    *    "eslint": "eslint"
    *  }
    */
    return {
        bin: npmWhich.sync(binName),
        args: files
    }
}
