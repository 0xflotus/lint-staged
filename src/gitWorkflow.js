const path = require('path')
const execa = require('execa')

module.exports = {

    getAbsolutePath(dir) {
        return path.isAbsolute(dir) ? dir : path.resolve(dir)
    },

    getCmdArgs(gitDir) {
        if (gitDir) {
            return ['--git-dir', this.getAbsolutePath(gitDir)]
        }
        return []
    },

    execGit(cmd, options) {
        const cwd = options && options.cwd ? options.cwd : process.cwd()
        const gitDir = options && options.gitDir
        return execa('git', this.getCmdArgs(gitDir).concat(cmd), { cwd: this.getAbsolutePath(cwd) })
    },

    gitStashSave(options) {
        return this.execGit(['stash', '--keep-index'], options)
    },

    gitStashPop(options) {
        return this.execGit(['stash'], options)
            .then(() => this.execGit(['stash', 'pop', 'stash@{1}'], options))
            .then(() => this.execGit(['read-tree', 'stash'], options))
            .then(() => this.execGit(['stash', 'drop'], options))
    },

    gitRestore(options) {
        return this.execGit(['checkout', '--merge', 'stash', '--', '.'], options)
            .then(() => this.execGit(['reset', '.'], options))
            .then(() => this.execGit(['stash', 'drop'], options))
    }
}
