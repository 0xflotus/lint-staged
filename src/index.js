var sgf = require('staged-git-files')
var minimatch = require('minimatch')
var ora = require('ora')
var assign = require('object-assign')
var appRoot = require('app-root-path')
var config = require(appRoot.resolve('package.json'))
var runScript = require('./runScript')

var defaultLinters = {}
var customLinters = config['lint-staged']
var linters = assign(defaultLinters, customLinters)
var spinner = ora('Starting lint-staged').start()

sgf('ACM', function (err, results) {
    if (err) {
        console.error(err)
    }
    var filePaths = results.map(function (file) {
        return file.filename
    })
    if (filePaths.length) {
        Object.keys(linters).forEach(function (key) {
            var linter = linters[key]
            var fileList = filePaths.filter(minimatch.filter(key, { matchBase: true }))
            if (fileList.length) {
                spinner.text = 'Running ' + linter + '...'
                runScript(linter, fileList, config, function (error, exitCode) {
                    if (error) {
                        console.error(error)
                    }
                    if (exitCode > 0) {
                        console.log('Linter %s exited with code %s', linter, exitCode)
                    }
                    spinner.stop()
                    spinner.clear()
                })
            } else {
                spinner.stop()
            }
        })
    } else {
        spinner.stop()
        console.log('No staged files found...')
    }
})

