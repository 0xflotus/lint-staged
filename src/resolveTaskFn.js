'use strict'

const dedent = require('dedent')
const execa = require('execa')
const symbols = require('log-symbols')
const findBin = require('./findBin')

const debug = require('debug')('lint-staged:task')

/**
 * Create and returns an error instance with given stdout and stderr. If we set
 * the message on the error instance, it gets logged multiple times(see #142).
 * So we set the actual error message in a private field and extract it later,
 * log only once.
 *
 * @param {string} linter
 * @param {string} errStdout
 * @param {string} errStderr
 * @returns {Error}
 */
function makeErr(linter, errStdout, errStderr) {
  const err = new Error()
  err.privateMsg = dedent`
    ${symbols.error} "${linter}" found some errors. Please fix them and try committing again.
    ${errStdout}
    ${errStderr}
  `
  return err
}

/**
 * Returns the task function for the linter. It handles chunking for file paths
 * if the OS is Windows.
 *
 * @param {Object} options
 * @param {string} options.linter
 * @param {string} options.gitDir
 * @param {Array<string>} options.pathsToLint
 * @param {number} options.maxPathsToLint
 * @returns {function(): Promise<string>}
 */
module.exports = function resolveTaskFn(options) {
  const { linter, gitDir, pathsToLint, maxPathsToLint } = options
  const { bin, args } = findBin(linter)

  const execaOptions = { reject: false }
  // Only use gitDir as CWD if we are using the git binary
  // e.g `npm` should run tasks in the actual CWD
  if (/git(\.exe)?$/i.test(bin) && gitDir !== process.cwd()) {
    execaOptions.cwd = gitDir
  }

  if (pathsToLint.length > maxPathsToLint) {
    return () =>
      Promise.reject(
        makeErr(
          linter,
          dedent`The file pattern specified for ${linter} matched ${pathsToLint.length} files.
            Attempting to run commands with too many arguments can result in cryptic
            errors on some platforms due to limitations on the total length of the command.
            Therefore, lint-staged will fail the task preemptively.

            It is recommended to run the linter on the entire project and commit the changes
            with --no-verify to bypass the pre-commit hook:

            \`git commit -m "Commit message" --no-verify\`

            Alternatively, the limit can be increased by specifying \`maxPathsToLint\`
            in the lint-staged config. See https://github.com/okonet/lint-staged#options.`,
          `Number of matched files exceeds limit(${maxPathsToLint}).`
        )
      )
  }

  return () => {
    // Execute the given linter binary with arguments and file paths using execa
    // and return the promise.
    const binArgs = args.concat(pathsToLint)

    debug('bin:', bin)
    debug('args: %O', binArgs)
    debug('opts: %o', execaOptions)

    return execa(bin, binArgs, execaOptions).then(result => {
      if (!result.failed) return `${symbols.success} ${linter} passed!`

      throw makeErr(linter, result.stdout, result.stderr)
    })
  }
}
