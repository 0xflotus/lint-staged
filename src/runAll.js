'use strict'

const sgf = require('staged-git-files')
const Listr = require('listr')
const has = require('lodash/has')
const pify = require('pify')
const exitHook = require('async-exit-hook')
const makeCmdTasks = require('./makeCmdTasks')
const generateTasks = require('./generateTasks')
const resolveGitDir = require('./resolveGitDir')
const git = require('./gitWorkflow')

const debug = require('debug')('lint-staged:run')

/**
 * Executes all tasks and either resolves or rejects the promise
 * @param config {Object}
 * @returns {Promise}
 */
module.exports = function runAll(config) {
  debug('Running all linter scripts')
  // Config validation
  if (!config || !has(config, 'concurrent') || !has(config, 'renderer')) {
    throw new Error('Invalid config provided to runAll! Use getConfig instead.')
  }

  const { concurrent, renderer, chunkSize, subTaskConcurrency } = config
  const gitDir = resolveGitDir()
  debug('Resolved git directory to be `%s`', gitDir)

  sgf.cwd = gitDir
  return pify(sgf)('ACM').then(files => {
    /* files is an Object{ filename: String, status: String } */
    const filenames = files.map(file => file.filename)
    debug('Loaded list of staged files in git:\n%O', filenames)

    const tasks = generateTasks(config, filenames).map(task => ({
      title: `Running tasks for ${task.pattern}`,
      task: () =>
        new Listr(
          makeCmdTasks(task.commands, task.fileList, {
            chunkSize,
            subTaskConcurrency
          }),
          {
            // In sub-tasks we don't want to run concurrently
            // and we want to abort on errors
            dateFormat: false,
            concurrent: false,
            exitOnError: true
          }
        ),
      skip: () => {
        if (task.fileList.length === 0) {
          return `No staged files match ${task.pattern}`
        }
        return false
      }
    }))

    const listrBaseOptions = {
      dateFormat: false,
      renderer
    }

    if (tasks.length) {
      return new Listr(
        [
          {
            title: 'Stashing changes...',
            enabled: () => filenames.length > 0,
            task: async (ctx, task) => {
              const hasPSF = await git.hasPartiallyStagedFiles()
              if (hasPSF) {
                ctx.hasStash = true
                await git.gitStashSave()
                // Restore working copy if we stashed but process has been terminated
                // TODO: Use tree shas to compare trees and determine if restore is needed
                exitHook(async () => {
                  await git.gitStashPop()
                })
              }
              return task.skip('No unstaged files found...')
            }
          },
          {
            title: 'Running linters...',
            task: () =>
              new Listr(
                tasks,
                Object.assign({}, listrBaseOptions, {
                  concurrent,
                  exitOnError: !concurrent // Wait for all errors when running concurrently
                })
              )
          },
          // Update index with hook fixes only if all linters pass
          {
            title: 'Updating index...',
            enabled: ctx => ctx.hasStash,
            skip: ctx => ctx.hasErrors && 'Skipping index update since there are errors',
            task: () => git.updateStash()
          },
          {
            title: 'Restoring local changes...',
            enabled: ctx => ctx.hasStash,
            task: () => git.gitStashPop()
          }
        ],
        listrBaseOptions
      ).run()
    }
    return 'No tasks to run.'
  })
}
