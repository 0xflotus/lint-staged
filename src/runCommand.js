var cp = require('child_process');
var npmWhich = require('npm-which')(process.cwd());

module.exports = function runLinter(linter, paths, cb) {
    npmWhich(linter, function(err, binPath) {
        var args = ['--'].concat(paths);
        if (err) {
            // Support for scripts from package.json
            binPath = 'npm'; // eslint-disable-line
            args = ['run', '-s', linter, '--'].concat(paths);
        }
        var npmStream = cp.spawn(binPath, args, {
            stdio: 'inherit' // <== IMPORTANT: use this option to inherit the parent's environment
        });
        npmStream.on('error', function(error) {
            process.exitCode = 1;
            cb.call(this, error, null);
        });
        npmStream.on('close', function(code) {
            if (typeof cb === 'function') {
                cb.call(this, null, code);
            }
        });
        npmStream.on('exit', function(code) {
            process.exitCode = code;
        });
    });
}
