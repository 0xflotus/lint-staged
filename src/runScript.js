var findBin = require('./findBin');
var spawn = require('child_process').spawn;

module.exports = function runScript(linter, paths, config, cb) {
    findBin(linter, paths, config, function(err, binPath, args) {
        if (err) {
            console.error(err);
        }
        var npmStream = spawn(binPath, args, {
            stdio: 'inherit' // <== IMPORTANT: use this option to inherit the parent's environment
        });
        npmStream.on('error', function(error) {
            process.exitCode = 1;
            cb.call(this, error, null);
        });
        npmStream.on('close', function(code) {
            if (typeof cb === 'function') {
                cb.call(this, null, code);
            }
        });
        npmStream.on('exit', function(code) {
            process.exitCode = code;
        });
    });
};
