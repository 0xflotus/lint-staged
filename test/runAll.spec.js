import makeConsoleMock from 'consolemock'
import sgfMock from 'staged-git-files'
import execa from 'execa'
import { getConfig } from '../src/getConfig'
import runAll from '../src/runAll'
import { saveStagedFiles, restoreStagedFiles, clearStagedFileStash } from '../src/gitWorkflow'

jest.mock('staged-git-files')
jest.mock('../src/gitWorkflow')

sgfMock.mockImplementation((params, callback) => {
  callback(null, [])
})
const globalConsoleTemp = global.console

describe('runAll', () => {
  beforeAll(() => {
    global.console = makeConsoleMock()
  })

  afterEach(() => {
    global.console.clearHistory()
    saveStagedFiles.mockClear()
    restoreStagedFiles.mockClear()
    clearStagedFileStash.mockClear()
  })

  afterAll(() => {
    global.console = globalConsoleTemp
  })

  it('should throw when invalid config is provided', async () => {
    await expect(runAll({})).rejects.toThrowErrorMatchingSnapshot()
    await expect(runAll()).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should not throw when a valid config is provided', () => {
    const config = getConfig({
      concurrent: false
    })
    expect(() => runAll(config)).not.toThrow()
  })

  it('should return a promise', () => {
    expect(runAll(getConfig({}))).toBeInstanceOf(Promise)
  })

  it('should resolve the promise with no tasks', async () => {
    expect.assertions(1)
    const res = await runAll(getConfig({}))

    expect(res).toEqual('No tasks to run.')
  })

  it('should resolve the promise with no files', async () => {
    expect.assertions(1)
    await runAll(getConfig({ linters: { '*.js': ['echo "sample"'] } }))
    expect(console.printHistory()).toMatchSnapshot()
  })

  it('should not skip tasks if there are files', async () => {
    expect.assertions(1)
    sgfMock.mockImplementationOnce((params, callback) => {
      callback(null, [{ filename: 'sample.js', status: 'sample' }])
    })
    await runAll(getConfig({ linters: { '*.js': ['echo "sample"'] } }))
    expect(console.printHistory()).toMatchSnapshot()
  })

  it('should not skip stashing and restoring if there are partially staged files', async () => {
    expect.assertions(4)
    sgfMock.mockImplementationOnce((params, callback) => {
      callback(null, [{ filename: 'sample.js', status: 'Modified' }])
    })
    await runAll(getConfig({ linters: { '*.js': ['echo "sample"'] } }))
    expect(saveStagedFiles).toHaveBeenCalledTimes(1)
    expect(restoreStagedFiles).toHaveBeenCalledTimes(0)
    expect(clearStagedFileStash).toHaveBeenCalledTimes(1)
    expect(console.printHistory()).toMatchSnapshot()
  })

  it('should skip updating stash if there are errors during linting', async () => {
    expect.assertions(3)
    sgfMock.mockImplementationOnce((params, callback) => {
      callback(null, [{ filename: 'sample.js', status: 'Modified' }])
    })
    execa.mockImplementation(() =>
      Promise.resolve({
        stdout: '',
        stderr: 'Linter finished with error',
        code: 1,
        failed: true,
        cmd: 'mock cmd'
      })
    )

    try {
      await runAll(getConfig({ linters: { '*.js': ['echo "sample"'] } }))
    } catch (err) {
      console.log(err)
    }
    expect(console.printHistory()).toMatchSnapshot()
    expect(saveStagedFiles).toHaveBeenCalledTimes(1)
    expect(restoreStagedFiles).toHaveBeenCalledTimes(1)
  })

  it('should skip linters and stash update but perform working copy restore if terminated', async () => {
    expect.assertions(3)
    sgfMock.mockImplementationOnce((params, callback) => {
      callback(null, [{ filename: 'sample.js', status: 'Modified' }])
    })
    execa.mockImplementation(() =>
      Promise.resolve({
        stdout: '',
        stderr: '',
        code: 0,
        failed: false,
        killed: true,
        signal: 'SIGINT',
        cmd: 'mock cmd'
      })
    )

    try {
      await runAll(getConfig({ linters: { '*.js': ['echo "sample"'] } }))
    } catch (err) {
      console.log(err)
    }
    expect(console.printHistory()).toMatchSnapshot()
    expect(saveStagedFiles).toHaveBeenCalledTimes(1)
    expect(restoreStagedFiles).toHaveBeenCalledTimes(1)
  })

  it('should reject promise when staged-git-files errors', async () => {
    expect.assertions(1)
    sgfMock.mockImplementationOnce((params, callback) => {
      callback('test', undefined)
    })

    try {
      await runAll(getConfig({}))
    } catch (err) {
      expect(err).toEqual('test')
    }
  })

  it('should skip stashing changes if no lint-staged files are changed', async () => {
    expect.assertions(3)
    sgfMock.mockImplementationOnce((params, callback) => {
      callback(null, [{ filename: 'sample.java', status: 'Modified' }])
    })
    execa.mockImplementationOnce(() =>
      Promise.resolve({
        stdout: '',
        stderr: 'Linter finished with error',
        code: 1,
        failed: true,
        cmd: 'mock cmd'
      })
    )

    try {
      await runAll(getConfig({ linters: { '*.js': ['echo "sample"'] } }))
    } catch (err) {
      console.log(err)
    }
    expect(console.printHistory()).toMatchSnapshot()
    expect(saveStagedFiles).toHaveBeenCalledTimes(0)
    expect(restoreStagedFiles).toHaveBeenCalledTimes(0)
  })
})
