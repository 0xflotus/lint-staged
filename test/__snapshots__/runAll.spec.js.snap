// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runAll should not skip stashing and restoring if there are partially staged files 1`] = `
"
LOG Backing up local changes... [started]
LOG Backing up local changes... [completed]
LOG Running linters... [started]
LOG Running tasks for *.js [started]
LOG echo \\"sample\\" [started]
LOG echo \\"sample\\" [completed]
LOG Running tasks for *.js [completed]
LOG Running linters... [completed]
LOG Clearing temporary backup... [started]
LOG Clearing temporary backup... [completed]"
`;

exports[`runAll should not skip tasks if there are files 1`] = `
"
LOG Backing up local changes... [started]
LOG Backing up local changes... [completed]
LOG Running linters... [started]
LOG Running tasks for *.js [started]
LOG echo \\"sample\\" [started]
LOG echo \\"sample\\" [completed]
LOG Running tasks for *.js [completed]
LOG Running linters... [completed]
LOG Clearing temporary backup... [started]
LOG Clearing temporary backup... [completed]"
`;

exports[`runAll should resolve the promise with no files 1`] = `
"
LOG No staged files match any of provided globs."
`;

exports[`runAll should skip linters and stash update but perform working copy restore if terminated 1`] = `
"
LOG Backing up local changes... [started]
LOG Backing up local changes... [completed]
LOG Running linters... [started]
LOG Running tasks for *.js [started]
LOG echo \\"sample\\" [started]
LOG echo \\"sample\\" [failed]
LOG → 
LOG Running tasks for *.js [failed]
LOG → 
LOG Running linters... [failed]
LOG Restoring local changes... [started]
LOG Restoring local changes... [completed]
LOG Clearing temporary backup... [started]
LOG Clearing temporary backup... [completed]
LOG {
  name: 'ListrError',
  errors: [
    {
      privateMsg: '\\\\n\\\\n\\\\n‼ echo \\"sample\\" was terminated with SIGINT',
      context: {hasErrors: true}
    }
  ],
  context: {hasErrors: true}
}"
`;

exports[`runAll should skip stashing changes if no lint-staged files are changed 1`] = `
"
LOG No staged files match any of provided globs."
`;

exports[`runAll should skip updating stash if there are errors during linting 1`] = `
"
LOG Backing up local changes... [started]
LOG Backing up local changes... [completed]
LOG Running linters... [started]
LOG Running tasks for *.js [started]
LOG echo \\"sample\\" [started]
LOG echo \\"sample\\" [failed]
LOG → 
LOG Running tasks for *.js [failed]
LOG → 
LOG Running linters... [failed]
LOG Restoring local changes... [started]
LOG Restoring local changes... [completed]
LOG Clearing temporary backup... [started]
LOG Clearing temporary backup... [completed]
LOG {
  name: 'ListrError',
  errors: [
    {
      privateMsg: '\\\\n\\\\n\\\\n× echo \\"sample\\" found some errors. Please fix them and try committing again.\\\\n\\\\nLinter finished with error',
      context: {hasErrors: true}
    }
  ],
  context: {hasErrors: true}
}"
`;

exports[`runAll should throw when invalid config is provided 1`] = `"Invalid config provided to runAll! Use getConfig instead."`;

exports[`runAll should throw when invalid config is provided 2`] = `"Invalid config provided to runAll! Use getConfig instead."`;
